Checksum Cleanup

schisamo opened the pull request:
**NOTE**: This PR is dependent on a FK `ON DELETE` constraint change represented by the following pull-request in `chef-sql-schema`: https://github.com/opscode/chef-sql-schema/pull/5

Properly cleaning up redundant checksums requires 2 distintive steps of related work:
### 1. Delete Redundant Checksums in Database
- **chef_sql:delete_cookbook_version/1 now cleans up checksums** - Attempt to delete all checksum records associated with a particular cookbook version relying on the fact the RDBMS will throw a foreign key constraint error _if_ the checksum record is still needed. YesSQL!
- **chef_db/delete_cookbook_version/2 should expose same contract** - This function needs to return the same values, namely the 1 or 2 indicating whether the enclosing cookbook was deleted.  This flag is needed for proper authz cleanup.
### 2. Delete Checksum Data in S3

This logic has been implemented in `chef_objects` and is represented in this pull-request:
https://github.com/opscode/chef_objects/pull/6

schisamo committed:
Add chef_sql:checsum_exists/1 function

schisamo committed:
add missing 'delete_checksum_by_id' mysql prepared statement

schisamo committed:
chef_sql:delete_cookbook_version/1 now cleans up checksums

Attempt to delete all checksum records associated with a particular
cookbook version relying on the fact the RDBMS will throw a foreign key
constraint error *if* the checksum record is still needed. YesSQL!

schisamo committed:
chef_db/delete_cookbook_version/2 should expose same contract

This function needs to return the same values, namely the 1 or 2
indicating whether the enclosing cookbook was deleted.  This flag is
needed for proper authz cleanup.

schisamo committed:
Add checksum deletion checks to integration tests

seth commented on priv/mysql_statements.config:
```diff
@@ -368,10 +368,16 @@
 {delete_sandbox_by_id,
  <<"DELETE FROM sandboxed_checksums WHERE sandbox_id = ?">>}.
 
+{find_checksum_by_id,
```
Since we don't share checksums across orgs, I think this query needs to take an org_id as well. Same for the delete_checksums_by_id query, needs to limit to a specific org_id.

christophermaier replied:
Yup

seth commented on src/chef_db.erl:
```diff
@@ -745,7 +746,13 @@ update_cookbook_version(#context{}=Ctx, UpdatedCookbookVersion, ActorId) ->
    -> {ok, 1 | 2} | not_found | {error, _}.
 delete_cookbook_version(#context{}=Ctx, #chef_cookbook_version{}=CookbookVersion) ->
-    delete_object(Ctx, delete_cookbook_version, CookbookVersion).
+    case delete_object(Ctx, delete_cookbook_version, CookbookVersion) of
```
The returns value of `delete_object` here is feeling increasingly leaky. I wonder if there is some smaller piece of delete_object to factor out for reuse? More on this later. Also, as-is I'm confused because it looks like delete_object isn't going to handle the 3-element return properly.

christophermaier replied:
The intention of bubbling up the deleted checksums was to separate S3/Bookshelf related operations (nuking the actual files) from the database operations (nuking our /knowledge/ of the files).  @schisamo and I talked about it, and this seemed like the most logical place to coordinate those activities.

christophermaier replied:
And yeah, delete_object looks like it needs a little 3-argument love.

seth commented on src/chef_sql.erl:
```diff
@@ -833,12 +835,12 @@ delete_cookbook_version(#chef_cookbook_version{id=CookbookVersionId,
                     case delete_cookbook_if_last(OrgId, Name) of
                         {ok, N} ->
-                            {ok, N + 1};
+                            {ok, N+1, DeletedChecksums};
```
Let's try to think through what things would look like with a more structured return value. This might be a record or a proplist like `{ok, [{cookbook_delete, true|false}, {deleted_checksums, Checksums}]}`.

Also, it might be worth seeing if refactoring some of this with a helper function that handles some of the branching in funheads might improve the readability. I think it would be easier to read with a bit less nesting.

christophermaier replied:
Maybe a small record... not sure I like a proplist for this, since you can't really pattern match on it.  I'm OK with the return value as it is, though (but I'd definitely want to refactor if more data were added later).

A de-nesting might be nice, but the nesting wasn't introduced here, so I'd be OK with postponing that particular refactor until later.

seth commented on src/chef_sql.erl:
```diff
@@ -888,9 +890,21 @@ create_sandbox(#chef_sandbox{} = Sandbox) ->
+                             boolean() | {error, term()}.
+checksum_exists(ChecksumId) ->
+    case sqerl:select(find_checksum_by_id, [ChecksumId], first_as_scalar, [checksum]) of
+        {ok, Checksum} when is_binary(Checksum) ->
```
minor style note: if you reorder the clauses you can remove the guard. Moreover, since you will only ever see Checksum as binary or none you could do:

```
{ok, Checksum} -> Checksum =/= none % or is_binary(Checksum)
```

to collapse the two clauses. It is fine as-is.

seth commented on src/chef_sql.erl:
```diff
@@ -1325,16 +1339,41 @@ fetch_cookbook_authz(OrgId, CookbookName) ->
                          [OrgId, CookbookVersionId]) of
         {ok, N} when is_integer(N) -> %% pretend there is 1
-            {ok, 1};
+            delete_checksums(Checksums);
```
Collapse these clauses since there is no difference in the follow on logic. So:

```
{ok, _} ->
    delete_checksums(Checksums)
```

seth commented on src/chef_sql.erl:
```diff
@@ -1325,16 +1339,41 @@ fetch_cookbook_authz(OrgId, CookbookName) ->
+                                {ok, N} when is_integer(N) -> %% pretend there is 1
+                                    [Checksum|Acc];
+                                {foreign_key, _} ->
+                                    %% The checksum may still be associated with
```
s/may/are/

seth commented on src/chef_sql.erl:
```diff
@@ -1325,16 +1339,41 @@ fetch_cookbook_authz(OrgId, CookbookName) ->
+                                    %% The checksum may still be associated with
+                                    %% another cookbook version record which is OK!
+                                    Acc;
+                                {error, _Error} ->
```
We should think about and add comments here on what we want to do if an error occurs. Why do you think swallowing the error is the right thing here? We should at least add some logging.

christophermaier replied:
Agree that comments and logging would be good.  I'm not sure what we could meaningfully do in an error situation, though, given that we've already deleted the cookbook version at this point, as well as (potentially) some number of checksums.  The horses have kind of left the barn at that point.

Now, if sqerl had transactions....

seth commented on src/chef_sql.erl:
```diff
@@ -1325,16 +1339,41 @@ fetch_cookbook_authz(OrgId, CookbookName) ->
+%% foreign_key constraint errors assuming the checksum is still associated with
+%% another cookbook_version_checksum record.  Returns a list of deleted checksum
+%% ids for further upstream processing(ie delete the checksums from S3).
+delete_checksums(Checksums) ->
```
Pretty sure this will need an OrgId arg.

christophermaier commented on src/chef_sql.erl:
```diff
@@ -888,9 +890,21 @@ create_sandbox(#chef_sandbox{} = Sandbox) ->
+%% @doc Helper function for testing checksum existence.
+-spec checksum_exists(Checksum :: binary()) ->
+                             boolean() | {error, term()}.
+checksum_exists(ChecksumId) ->
```
It looks like this function is just required for testing, and isn't used in the actual operating code, right?

In that case, it would be a good idea to pull it into the itests file or (better) pull out a separate testing helper module and stick it there (since the itest file is getting ridiculously large).  You can add the prepared statement into Sqerl when the tests are set up (see what I did with the cookbook test setup)

christophermaier commented on src/chef_sql.erl:
```diff
@@ -1325,16 +1339,41 @@ fetch_cookbook_authz(OrgId, CookbookName) ->
+                      end,
+                      [],
+                      Checksums),
+    {ok, lists:reverse(DeletedChecksums)}.
```
I don't think list reversal is needed here, given that checksums don't have any inherent order.

schisamo committed:
checksum queries should be scoped by by `org_id`

schisamo committed:
delete_checksums/2 -> delete_cookbook_version_checksums/2

The new name better reflects the actual purpose of the function.  A 
version of chef_sql:delete_checksums/2 still exists whose job is to 
actually delete checksum records.

schisamo committed:
remove frivolous list reversal

schisamo committed:
collapse match clauses..yeah code removal

schisamo committed:
demote chef_sql:checksum_exists/2 to an itest helper

This function is required purely for testing, not actual operating code.

schisamo committed:
log any errors that occur while deleting checksum records.

schisamo commented:
^^ @christophermaier @seth 

I still need to sort out the return record from `chef_sql:delete_cookbook_version/1` but I made some progress based on your feedback.

christophermaier commented:
So far, so good!

schisamo committed:
chef_sql/delete_cookbook_version/1 now returns an actual record

schisamo committed:
update chef_db:delete_cookbook_version/2 to use new record

chef_sql:delete_cookbook_version now returns a chef_db_cb_version_delete
record.

schisamo committed:
simplify chef_db:delete_object/3

We can remove the #chef_cookbook_version{} special case of delete_object
and collapse the return matching logic.

schisamo committed:
update integration tests for new record return value

Add the local project 'include' dir to the erlang include path for all
erlang commands in the itest_dun target.

schisamo committed:
fix dialyzer errors

* assorted spec changes
* chef_db:get_id/1 missing a func head for chef_cookbook_version

schisamo commented:
^^ @christophermaier @seth got the new record worked in.  Ratholed slightly fixing numerous dialyzer errors.  Will be starting on the actual s3 deletion logic in the AM.

schisamo committed:
wire in chef_s3:delete_checksums/2 for checksum cleanup on S3

schisamo commented:
^^ @seth @christophermaier final feedback when you have a sec.

seth commented on itest/itest.erl:
```diff
@@ -1598,32 +1602,69 @@ delete_cookbook_version_checksums() ->
-                 chef_sql:delete_cookbook_version(Got)),
-    %% TODO - how to check if the checksums have been deleted ?
+
+    % Verify all checksums exist
```
Minor, but prefer `%%` for line comments

seth commented on src/chef_sql.erl:
```diff
@@ -1322,18 +1326,45 @@ fetch_cookbook_authz(OrgId, CookbookName) ->
+                    %% another cookbook version record which is OK!
+                    Acc;
+                {error, Error} ->
+                    error_logger:error_msg("The following error occured
```
Since this is an internal error message, I'd shorten this to indicate the Mod and Fun and details specific to the error.

seth commented on src/chef_sql.erl:
```diff
@@ -832,13 +834,16 @@ update_cookbook_version_checksums(#chef_cookbook_version{ id        = Id,
+                            DeleteResponse#chef_db_cb_version_delete{cookbook_delete = false};
+                        {ok, 1} ->
+                            DeleteResponse#chef_db_cb_version_delete{cookbook_delete = true};
                         Error = {error, _} ->
```
This is ok, but I think might be easier to read with a helper function and not as a many "exits". For example:

```
delete_cookbook_version(#chef_cookbook_version{id=CookbookVersionId,
                                               org_id=OrgId,
                                               name=Name}) ->
    case delete_cookbook_version_checksums(OrgId, CookbookVersionId) of
        {ok, DeletedChecksums} ->
            CookbookDelete = case delete_object(delete_cookbook_version_by_id, CookbookVersionId) of
                                 {ok, 1} ->
                                     case delete_cookbook_if_last(OrgId, Name) of
                                         {ok, 0} -> false;
                                         {ok, 1} -> true;
                                         Error = {error, _} -> Error
                                     end;
                                 {error, Reason} ->
                                     {error, Reason}
                             end,
            dcv_result_or_error(CookbookDelete, DeletedChecksums);
        {error, Reason} ->
            {error, Reason}
    end.

dcv_result_or_error({error, Reason}, _) ->
    {error, Reason};
dcv_result_or_error(CookbookDelete, DeletedChecksums) ->
    #chef_db_cb_version_delete{deleted_checksums = DeletedChecksums,
                               cookbook_delete = CookbookDelete}.
```

schisamo committed:
Use helper function to improve delete_cookbook_version/1 readability

schisamo committed:
Fix comment: 2 % chars plz

schisamo committed:
Use an internal log message.

schisamo committed:
remove the Gemfile.lock on itest_clean

This ensures we have latest version of of the schema every itest run.

schisamo committed:
re-add special-case of chef_db:delete_object/3 for cookbook versions

In order to delete a cookbook version we need an OrgId and Name in 
addition to the Id.

schisamo commented:
This has already been merged.
